internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,308 @@
+// +kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadmindownloadrequests/finalizers,verbs=update
+// +kubebuilder:rbac:groups=velero.io,resources=downloadrequests,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=velero.io,resources=downloadrequests/status,verbs=get
+
+// Reconcile the NonAdminDownloadRequest object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// This reconcile implements ObjectReconciler https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile#ObjectReconciler
+// Each reconciliation event gets the associated object from Kubernetes before passing it to Reconcile
+// Notes for test, this function expect req to come in with k8s UID already populated like real cluster
+func (r *NonAdminDownloadRequestReconciler) Reconcile(ctx context.Context, req *nacv1alpha1.NonAdminDownloadRequest) (reconcile.Result, error) {
+	if req == nil || req.Spec.Target.Kind == constant.EmptyString {
+		return ctrl.Result{}, nil
+	}
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling NonAdminDownloadRequest", "name", req.Name, "namespace", req.Namespace)
+	// defines associated downloadrequest for getting, or deleting
+	veleroDR := velerov1.DownloadRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.VeleroDownloadRequestName(),
+			Namespace: r.OADPNamespace,
+			Labels: func() map[string]string {
+				nal := function.GetNonAdminLabels()
+				nal[constant.NadrOriginNACUUIDLabel] = string(req.GetUID())
+				return nal
+			}(),
+			Annotations: function.GetNonAdminDownloadRequestAnnotations(req),
+		},
+		Spec: velerov1.DownloadRequestSpec{
+			Target: velerov1.DownloadTarget{
+				Kind: req.Spec.Target.Kind,
+			},
+		},
+	}
+	// if request has status, it may already be processed or expired
+	if req.Status.VeleroDownloadRequest.Status != nil {
+		// if request is expired, delete NADR after deleting velero DR
+		if req.Status.VeleroDownloadRequest.Status.Expiration != nil && req.Status.VeleroDownloadRequest.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+			// find associated velero downloadrequest and delete that first
+			return reconcile.Result{}, r.deleteExpiredDownloadRequest(ctx, &veleroDR, req)
+		}
+		// if request is not expired, and has downloadUrl, requeue when expired to cleanup.
+		if req.Status.VeleroDownloadRequest.Status.DownloadURL != constant.EmptyString {
+			return reconcile.Result{RequeueAfter: time.Until(req.Status.VeleroDownloadRequest.Status.Expiration.Time)}, nil
+		}
+	}
+	// try get veleroDR if exists, then update status
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroDR.Namespace, Name: veleroDR.Name}, &veleroDR); err == nil {
+		return reconcile.Result{}, r.updateNADRWithURL(ctx, &veleroDR, req)
+	} else if !apierrors.IsNotFound(err) {
+		// some other errors, requeue to retry get
+		return reconcile.Result{}, err
```

Comment:
I think no error here is terminal, like creating a NADR for a non existent NAB. Should some errors be terminal?

Reply:
Up to team.

Reply:
mimic upstream.

Reply:
```
		if err := r.client.Get(ctx, kbclient.ObjectKey{
			Namespace: downloadRequest.Namespace,
			Name:      backupName,
		}, backup); err != nil {
			if apierrors.IsNotFound(err) {
				log.WithError(err).Error("fail to get backup for DownloadRequest")
				return ctrl.Result{}, nil
			}
			log.Warnf("fail to get backup for DownloadRequest %s. Retry later.", err.Error())
			return ctrl.Result{}, errors.WithStack(err)
```

[velero retry until backup exists](https://github.com/vmware-tanzu/velero/blob/main/pkg/controller/download_request_controller.go#L156C1-L165C47).. so this matches our retry. nothing to do. resolving.